lib/pretty-repl.js

```diff
@@ -17,47 +62,113 @@ class PrettyREPLServer extends repl.REPLServer {
+      // First, we simplify whatever existing line structure is present in a
+      // way that preserves highlighting of any subsequent part of the code.
+      // The goal here is to reduce the amount of code that needs to be highlighted,
+      // because this typically runs once for each character that is entered.
+      const simplified = this._stripCompleteJSStructures(this.lineBeforeInsert);
+
+      // Colorize the 'before' state.
+      const before = this._doColorize(simplified);
+      // Colorize the 'after' state, using the same simplification (this works because
+      // `lineBeforeInsert + stringToWrite === line` implies that
+      // `simplified       + stringToWrite` is a valid simplification of `line`.
+      const after = this._doColorize(simplified + stringToWrite);
+
+      // Find the first character or escape sequence that differs in `before` and `after`.
+      const commonPrefixLength = computeCommonPrefixLength(before, after);
+
+      // Gather all escape sequences that occur in the *common* part of the string.
+      // Applying them all one after another puts the terminal into the state of the
+      // highlighting at the divergence point.
+      // (This makes the assumption that those escape sequences only affect formatting,
+      // not e.g. cursor position, which seem like a reasonable assumption to make
+      // for the output from a syntax highlighter).
+      let ansiStatements = (before.slice(0, commonPrefixLength).match(ansiRegexMatchAll) || []);
+
+      // Filter out any foreground color settings before the last reset (\x1b[39m).
+      // This helps reduce the amount of useless clutter we write a bit, and in
+      // particular helps the mongosh test suite not ReDOS itself when verifying
+      // output coloring.
+      const lastForegroundColorReset = ansiStatements.lastIndexOf('\x1b[39m');
+      if (lastForegroundColorReset !== -1) {
+        ansiStatements = ansiStatements.filter((sequence, index) => {
+          // Keep escape sequences that come after the last full reset or modify
+          // something other than the foreground color.
+          return index > lastForegroundColorReset || !sequence.match(/^\x1b\[3\d.*m$/);
+        });
+      }
+
+      // In order to get from `before` to `after`, we have to reduce the `before` state
+      // back to the common prefix of the two. Do that by counting all the
+      // non-escape-sequence characters in what comes after the common prefix
+      // in `before`.
+      const backtrackLength = characterCount(stripAnsi(before.slice(commonPrefixLength)));
+
+      // Put it all together: Backtrack from `before` to the common prefix, apply
+      // all the escape sequences that were present before, and then apply the
+      // new output from `after`.
+      this.output.write('\b'.repeat(backtrackLength) + ansiStatements.join('') + after.slice(commonPrefixLength));
+    } else if (stringToWrite.startsWith(this._prompt)) {
```

Comment:
Looking at this if all the branches seems to have no interactions, maybe is possible to extract it to 3 functions with early return, make it more easy to parse and less intimidating to work on, something like:

```
if (this.lineBeforeInsert !== undefined && this.lineBeforeInsert + stringToWrite === this.line) {
   return writeAppendedString(...);
}

if (stringToWrite.startsWith(this._prompt)) {
    return writeFullLine();
}

return super._writeToOutput(stringToWrite);
```

Reply:
Okay, sure – I’ve done that and adjusted the comments a bit here for that, because the content of the conditions is related to the contents of the blocks/methods for each case